packages/babel-parser/src/parser/expression.js

```diff
@@ -704,6 +695,73 @@ export default class ExpressionParser extends LValParser {
     }
   }
 
+  checkSliceArgument(pos: number, expr: N.Expression) {
+    if (
+      expr.type === "NumericLiteral" ||
+      expr.type === "Identifier" ||
+      (expr.type === "UnaryExpression" &&
+        expr.operator === "-" &&
+        (expr.argument.type === "NumericLiteral" ||
+          expr.argument.type === "Identifier"))
+    ) {
+      return;
+    }
+    this.raise(
+      pos,
+      "Slice argument requires to be either numeric literal or identifier",
+    );
+  }
+
+  parseSliceArgument(): ?N.Expression {
+    let expr = null;
+    this.eat(tt.colon);
+    const pos = this.state.pos;
+    if (!this.match(tt.colon) && !this.match(tt.bracketR)) {
+      expr = this.parseExpression();
+      this.checkSliceArgument(pos, expr);
+    }
+    return expr;
+  }
+
+  parseMaybeSliceExpression(
+    startPos: number,
+    startLoc: Position,
+    base: N.Expression,
+    state: N.ParseSubscriptState,
+  ): N.Expression {
+    const node = this.startNodeAt(startPos, startLoc);
+    node.object = base;
+    let expr;
+    const pos = this.state.pos;
+    if (!this.hasPlugin("sliceNotation") || !this.match(tt.colon)) {
```

Comment:
This is kinda complicated to follow.

Reply:
Yeah I admit it looks complicated. The code was basically rewritten from
```js
if (this.hasPlugin("sliceNotation")) {
  if (!this.match(tt.colon)) {
    expr = this.parseExpression();
  } else {
    // SliceExpression stuffs.
  }
  if (!this.match(tt.colon)) {
    // MemberExpression stuffs
  } else {
    // SliceExpression stuffs.
  }
} else {
  expr = this.parseExpression();
  // MemberExpression stuffs
}
```

Maybe I could extract `// MemberExpression stuffs` into a dedicate function after I come up with a proper name.